### PR TITLE
refactor: update request-id recipe to use contextvars

### PR DESCRIPTION
**Pull request by EricGoulart**
# Summary of Changes

Migrate the request-id handling from threading.local() to contextvars to improve compatibility with async coroutines and avoid issues with threading. This change ensures that request-id is properly tracked in asynchronous environments, providing more robust handling in both sync and async contexts.
Previously, threading.local() was used, which does not handle coroutines effectively. By using contextvars, we ensure that the request-id remains consistent across async calls.

# Related Issues

Closes https://github.com/falconry/falcon/issues/2260.

### CODE REVIEW

**Comment by EricGoulart** [comment]
I wanted to apologize for the confusion with my pull request. This is my first time contributing to a public repository, and I made a mistake with the process. I had previously opened a PR with the changes, but I accidentally closed it and am unable to reopen it.

File: tests/test_recipes.py [hunk 1]
```diff
@@ -141,3 +141,42 @@ def test_raw_path(self, asgi, app_kind, util):
...
+        request_id2 = response.json['request_id']
+
+        assert request_id1 != request_id2
+
+    def test_request_id_in_response_header(self, app):
+        client = falcon.testing.TestClient(app)
+
+        response = client.simulate_get('/test')
+        assert 'X-Request-ID' in response.headers
+        assert response.headers['X-Request-ID'] == response.json['request_id']
```
Comment (CaselIT) [inline tests/test_recipes.py:182]: could also test that's the same as the value returned in the body

**Review by CaselIT — DISMISSED** [review]

File: docs/user/recipes/request-id.rst [hunk 2]
```diff
@@ -10,14 +10,14 @@ to every log entry.
 
 If you wish to trace each request throughout your application, including
 from within components that are deeply nested or otherwise live outside of the
-normal request context, you can use a `thread-local`_ context object to store
+normal request context, you can use a `contextvars`_ object to store
```
Comment (vytas7) [inline docs/user/recipes/request-id.rst:13]: We could simply replace this with an Intersphinx link to the `contextvars` module, but this applies to the current version too, so not a requirement for merging this.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Thanks, this looks good now at first glance. :+1:

However, many CI gates are still failing, could you please take a look?
(To save some time on review and CI rounds, it may be more efficient to run `tox` locally first on your side, and make sure everything is green.)

**Review by vytas7 — APPROVED** [review]
I fixed the import chain issues in the unit tests, thanks once again for this improvement @EricGoulart!

(I'll address two minor docs things remaining in a follow-up PR.)

**Pull request merged**